Using custom exception class for network loading error

matanshukry opened the pull request:
## Description

Created a custom exception class for use.

## Related Issues

#35333 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

p.s.
The implementation of the `exception.toString` method will output the message directly so that this won't be a breaking change for anyone who is dependent on the syntax of the message. Let me know if this is the right way to go.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

matanshukry committed:
Using custom exception class for network loading error

chunhtai commented on packages/flutter/lib/src/painting/_network_image_io.dart:
```diff
@@ -122,3 +127,28 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
+
+/// Thrown to indicate that loading a network image failed
+///
+class NetworkImageLoadException implements Exception {
```
It feel weird to have Exception defined here, and given the information you need, it can be more general for http usage maybe something like HttpException. and I think it should be implemented in [dart sdk](https://github.com/dart-lang/sdk/blob/bbb027aa2ac4ebbb0023227347c3e2d33661b80e/sdk/lib/core/exceptions.dart#L20)
@dnfield do you have any input on this?

dnfield replied:
Can we just use https://api.dartlang.org/stable/2.4.0/dart-io/HttpException-class.html?

dnfield replied:
We don't get the status code in that case, but maybe we coulf file a bug upstream for that if it's important.

matanshukry replied:
Agreed, created an issue there. I'll probably do the PR there as well.
https://github.com/dart-lang/sdk/issues/37423

Can we put this with label blocked until the other issue is reolved?

dnfield commented on packages/flutter/lib/src/painting/_network_image_io.dart:
```diff
@@ -15,13 +15,15 @@ import 'image_provider.dart' as image_provider;
 
 /// The dart:io implemenation of [image_provider.NetworkImage].
-class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkImage> implements image_provider.NetworkImage {
+class NetworkImage
```
Please revert all of the formatting related changes.

The Flutter repository does not use dartfmt - see our style guide for more information (https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo)

matanshukry replied:
Fixed. Check it again?

matanshukry committed:
Reverted auto-format with dart-sdk

goderbauer commented:
(PR triage): Can we close this PR for now until the upstream changes are in?

matanshukry commented:
@goderbauer by upstream changes you mean the issue in Dart?

And why do you want to close it? The issue is gonna stay open anyway..

dnfield commented:
@matanshukry it helps us when we're trying to keep track of what actually needs attention.  This PR isn't really actionable at the moment until the Dart issue is fixed.  We should leave the issue open for sure, but if we can close this in the mean time it will help us make sure we're paying attention to pull requests that are actionable.

dnfield commented:
And of course, once the upstream issue is resolved we can re-open this

matanshukry commented:
@dnfield k I guess. What about the failed checks? I couldn't really figure out if it was something with my code I should fix/do?

dnfield commented:
Almost certainly a flake.  It would likely pass on a rerun.

We can rerun them again when we reopen this.

matanshukry commented:
@dnfield Dart team doesn't like the idea of using status code. Look up https://github.com/dart-lang/sdk/issues/37423 , latest response.

What do you recommend then? creating an HttpException in flutter (with status code), or using the NetworkImageLoadException?

dnfield commented on packages/flutter/lib/src/painting/_network_image_io.dart:
```diff
@@ -122,3 +122,25 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
 }
+
+/// Thrown to indicate that loading a network image failed
+///
```
Nits: remove blank line, and make the first line a complete sentence ending with a `.`.  Perhaps something like `The exception thrown when the HTTP request to load a network image fails.`

jonahwilliams commented on packages/flutter/lib/src/painting/_network_image_io.dart:
```diff
@@ -122,3 +122,25 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
+  final int code;
+
+  /// A human-readable error message.
+  final String message;
```
This doesn't need to be public.

dnfield commented on packages/flutter/lib/src/painting/_network_image_io.dart:
```diff
@@ -122,3 +122,25 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
+///
+class NetworkImageLoadException implements Exception {
+  /// Creates a [NetworkImageLoadException] with the specified http status
+  /// [code] and the [uri]
```
nit: end with `.`.

We should document here that `uri` must not be null.  Should also either remove `@required` from `code`, or `assert(code != null)`.

dnfield replied:
I would go with `assert(code != null)`

dnfield replied:
This one may still need attention - and now `[code]` should be updated to `[statusCode]`.

Thanks for bearing with us on this!

jonahwilliams commented on packages/flutter/lib/src/painting/_network_image_io.dart:
```diff
@@ -122,3 +122,25 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
+    : assert(uri != null),
+      message = 'HTTP request failed, statusCode: $code, $uri';
+
+  /// An http status code.
```
This should describe the relationship of this status code to the request, and also if it is nullable.

dnfield replied:
Perhaps `The HTTP status code from the server.`

jonahwilliams commented on packages/flutter/lib/src/painting/_network_image_io.dart:
```diff
@@ -122,3 +122,25 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
+      message = 'HTTP request failed, statusCode: $code, $uri';
+
+  /// An http status code.
+  final int code;
```
Call this `statusCode`

jonahwilliams commented on packages/flutter/lib/src/painting/_network_image_io.dart:
```diff
@@ -122,3 +122,25 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
+
+  @override
+  String toString() => message;
+}
```
Add anewline at EOF.

dnfield replied:
nit: Please end files with a blank line

dnfield commented on packages/flutter/lib/src/painting/_network_image_io.dart:
```diff
@@ -87,7 +87,7 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
       final HttpClientResponse response = await request.close();
       if (response.statusCode != HttpStatus.ok)
-        throw Exception('HTTP request failed, statusCode: ${response?.statusCode}, $resolved');
+        throw NetworkImageLoadException(code: response?.statusCode, uri: resolved);
```
This should fail analysis because `code` is `@required`. 

But we already know that `response` is not null, so this should just be `response.statusCode` (and I realize this is a problem in the existing code :).

jonahwilliams commented:
This needs a test that asserts that the exception is thrown correctly, and contains the expected toString for a particular request and statusCode.

dnfield reviewed (changes requested):
This will need tests.  For example, you could write a test that mocks out the request/response to return an error code, and validates that the correct exception is thrown with the expected status code.

matanshukry committed:
Review changes

matanshukry commented:
I made all the changes except the tests. I'll probably create a test tomorrow. Let me know about the other changes for now. There were a lot of comments (Which is awesome btw!), but I'm pretty sure I didn't miss anything.

matanshukry committed:
Added test

matanshukry commented:
Added test.

p.s.
I've added the line `export '_network_image_io.dart';` so that the exception class `NetworkImageLoadException` will be visible, although I'm not sure that's the right way to go. Let me know if there's a better way you would like me to handle it.

matanshukry committed:
Fixed analyzer errors

jonahwilliams commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -17,6 +17,8 @@ import 'binding.dart';
 import 'image_cache.dart';
 import 'image_stream.dart';
 
+export '_network_image_io.dart';
```
Instead of exporting here for the network image, define the exception class in image_provider.dart

matanshukry replied:
Done

jonahwilliams replied:
This line needs to be removedstill

matanshukry replied:
Yeah forgot about it, sorry. Fixed now.

matanshukry committed:
Moved exception class

dnfield reviewed (approved):
LGTM!

jonahwilliams reviewed (dismissed):
Please remove the exported private library

matanshukry committed:
Removed the unneeded export

matanshukry commented:
> Please remove the exported private library

Sorry forgot about that. Fixed now.

chunhtai reviewed (approved):
LGTM

matanshukry commented:
Fro reading the guide about contributing, now that I got mr LGTM (yay :P) this PR just needs to be merged by one of the contributors with a merge access, right?

I do see 2 post-commit tests have failed, not sure if I'm suppose to do something about it though?

dnfield commented:
I rekicked the test.  We can land on green

Piinks commented:
Kicked `integration_tests-windows` again. If it times out again, I would recommend updating the branch.
